src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1006,251 +1047,375 @@ RouterSelectTask(Query *originalQuery, Query *query)
 	task->upsertQuery = upsertQuery;
 	task->requiresMasterEvaluation = false;
 
+	*placementList = workerList;
+
 	return task;
 }
 
 
 /*
- * RouterQueryJob creates a Job for the specified query to execute the
- * provided single shard select task.
+ * TargetShardIntervalsForSelect performs shard pruning for all referenced relations
+ * in the query and returns list of shards per relation. Shard pruning is done based
+ * on provided restriction context per relation. The function bails out and returns NULL
+ * if any of the relations has no active shards. It also records pruned shard intervals
+ * in relation restriction context to be used later on.
  */
-static Job *
-RouterQueryJob(Query *query, Task *task)
+static List *
+TargetShardIntervalsForSelect(Query *query,
+							  RelationRestrictionContext *restrictionContext)
 {
-	Job *job = NULL;
-	List *taskList = NIL;
-	TaskType taskType = task->taskType;
+	List *prunedRelationShardList = NIL;
+	ListCell *restrictionCell = NULL;
 
-	/*
-	 * We send modify task to the first replica, otherwise we choose the target shard
-	 * according to task assignment policy.
-	 */
-	if (taskType == MODIFY_TASK)
-	{
-		taskList = FirstReplicaAssignTaskList(list_make1(task));
-	}
-	else
+	Assert(query->commandType == CMD_SELECT);
+	Assert(restrictionContext != NULL);
+
+	foreach(restrictionCell, restrictionContext->relationRestrictionList)
 	{
-		taskList = AssignAnchorShardTaskList(list_make1(task));
-	}
+		RelationRestriction *relationResriction =
+			(RelationRestriction *) lfirst(restrictionCell);
+		Oid relationId = relationResriction->relationId;
+		Index tableId = relationResriction->index;
+		DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(relationId);
+		int shardCount = cacheEntry->shardIntervalArrayLength;
+		List *baseRestrictionList = relationResriction->relOptInfo->baserestrictinfo;
+		List *restrictClauseList = get_all_actual_clauses(baseRestrictionList);
+		List *shardIntervalList = NIL;
+		List *prunedShardList = NIL;
 
-	job = CitusMakeNode(Job);
-	job->dependedJobList = NIL;
-	job->jobId = INVALID_JOB_ID;
-	job->subqueryPushdown = false;
-	job->jobQuery = query;
-	job->taskList = taskList;
+		relationResriction->prunedShardIntervalList = NIL;
 
-	return job;
+		/* error out if no shards exist for any of the relations */
+		if (shardCount == 0)
+		{
+			return NULL;
+		}
+
+		shardIntervalList = LoadShardIntervalList(relationId);
```

Comment:
Yes and no.

PruneShardList expects a list of shards. I used ```cacheEntry->sortedShardIntervalArray``` to create a list and send it over for pruning.